diff -up qtbase-opensource-src-5.9.1/src/plugins/sqldrivers/mysql/qsql_mysql.cpp.0502 qtbase-opensource-src-5.9.1/src/plugins/sqldrivers/mysql/qsql_mysql.cpp
--- qtbase-opensource-src-5.9.1/src/plugins/sqldrivers/mysql/qsql_mysql.cpp.0502	2017-06-28 04:54:29.000000000 -0500
+++ qtbase-opensource-src-5.9.1/src/plugins/sqldrivers/mysql/qsql_mysql.cpp	2017-09-27 09:40:49.397425236 -0500
@@ -1158,16 +1158,22 @@ static void qLibraryInit()
     }
 # endif // MYSQL_VERSION_ID
 #endif // Q_NO_MYSQL_EMBEDDED
+
+#ifdef MARIADB_BASE_VERSION || defined(MARIADB_VERSION_ID)
+    qAddPostRoutine(mysql_server_end);
+#endif
 }
 
 static void qLibraryEnd()
 {
-#ifndef Q_NO_MYSQL_EMBEDDED
-# if MYSQL_VERSION_ID > 40000
-#  if (MYSQL_VERSION_ID >= 40110 && MYSQL_VERSION_ID < 50000) || MYSQL_VERSION_ID >= 50003
-    mysql_library_end();
-#  else
-    mysql_server_end();
+#if !defined(MARIADB_BASE_VERSION) || !defined(MARIADB_VERSION_ID)
+# if !defined(Q_NO_MYSQL_EMBEDDED)
+#  if MYSQL_VERSION_ID > 40000
+#   if (MYSQL_VERSION_ID >= 40110 && MYSQL_VERSION_ID < 50000) || MYSQL_VERSION_ID >= 50003
+     mysql_library_end();
+#   else
+     mysql_server_end();
+#   endif
 #  endif
 # endif
 #endif
