diff -up qtbase-opensource-src-5.6.0/config.tests/unix/journald/journald.pro.systemd qtbase-opensource-src-5.6.0/config.tests/unix/journald/journald.pro
--- qtbase-opensource-src-5.6.0/config.tests/unix/journald/journald.pro.systemd	2016-02-25 10:25:37.000000000 -0600
+++ qtbase-opensource-src-5.6.0/config.tests/unix/journald/journald.pro	2016-06-03 21:41:09.119559127 -0500
@@ -1,6 +1,6 @@
 SOURCES = journald.c
 
 CONFIG += link_pkgconfig
-PKGCONFIG_PRIVATE += libsystemd-journal
+PKGCONFIG_PRIVATE += libsystemd
 
 CONFIG -= qt
diff -up qtbase-opensource-src-5.6.0/src/corelib/global/global.pri.systemd qtbase-opensource-src-5.6.0/src/corelib/global/global.pri
--- qtbase-opensource-src-5.6.0/src/corelib/global/global.pri.systemd	2016-02-25 10:25:37.000000000 -0600
+++ qtbase-opensource-src-5.6.0/src/corelib/global/global.pri	2016-06-04 01:46:06.714935329 -0500
@@ -53,7 +53,7 @@ slog2 {
 
 journald {
     CONFIG += link_pkgconfig
-    PKGCONFIG_PRIVATE += libsystemd-journal
+    PKGCONFIG_PRIVATE += libsystemd
     DEFINES += QT_USE_JOURNALD
 }
 
